support timedelta in Q_.init

andrewgsavage opened the pull request:
- [x] Closes #608 #690
- [x] Executed `pre-commit run --all-files` with no errors
- [x] The change is fully covered by automated unit tests
- [ ] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

andrewgsavage committed:
support timedelta in Q_.init

andrewgsavage committed:
changes

andrewgsavage committed:
numpy

keewis commented on pint/compat.py:
```diff
@@ -293,6 +298,26 @@ def is_duck_array_type(cls: type) -> bool:
+    if isinstance(obj, datetime.timedelta):
+        return obj.total_seconds()
+    elif isinstance(obj, np_timedelta64) or obj.dtype == np_timedelta64:
+        return obj.astype(float)
```
this is not going to work in general: `timedelta64` has its own units, which can range from days (or years?) to nanoseconds, and `astype(float)` is the equivalent of `magnitude`. I don't have any advice on how to extract that, though, other than parsing the string repr of the dtype.

keewis reviewed (commented):
I've got a couple of comments, mostly regarding `timedelta64` and where to put the code handling it.

I would also like to have a way to convert a quantity with dimension `[time]` to a `timedelta64`, but that can be its own PR (the API is not quite clear to me: should `to` special-case units of `timedelta64[us]`, for example?).

andrewgsavage committed:
td64 dtypes that aren't second

jakirkham commented:
It looks like this is ready for another review

andrewgsavage committed:
Merge branch 'master' into infer_time

andrewgsavage commented:
poke @keewis

andrewgsavage committed:
Merge branch 'master' into infer_time

andrewgsavage committed:
fix

andrewgsavage commented:
> I would also like to have a way to convert a quantity with dimension `[time]` to a `timedelta64`, but that can be its own PR (the API is not quite clear to me: should `to` special-case units of `timedelta64[us]`, for example?).

I think to do we should make time into a nonmultiplactive unit like temperature, so there is then a delta_time unit
the new c++ library does this for all units with two different classes, a point and a vector
https://github.com/hgrecco/pint/issues/2120

keewis reviewed (commented):
sorry for disappearing for quite a while.

Looks good to me in general, but I have two comments.

andrewgsavage committed:
Merge branch 'master' into infer_time

andrewgsavage committed:
use facets

andrewgsavage committed:
factes

andrewgsavage committed:
Update pint/facets/numpy/quantity.py

Co-authored-by: Justus Magin <keewis@users.noreply.github.com>

andrewgsavage committed:
run numpy tests

andrewgsavage committed:
Move datetime.timedelta handling into PlainQuantity

Per keewis's review, _is_timedelta/_convert_timedelta moved into the
numpy facet to keep np.timedelta64-specific detection out of plain.
But that also pulled in the stdlib datetime.timedelta branch, so
PlainQuantity.__new__ ended up depending on a NumpyQuantity-only
method (only working because NumpyQuantity is unconditionally mixed
into the default Quantity class).

datetime.timedelta needs nothing numpy-related, so give PlainQuantity
its own _is_timedelta/_convert_timedelta handling it directly, and
have NumpyQuantity override both to add the np.timedelta64/duck-array
cases, delegating to super() for the plain datetime.timedelta case.

Also switch the __new__ call site from the unbound-style
cls._is_timedelta(cls, value) to a normal inst._is_timedelta(value)
call now that it's a real instance method on the class being
constructed.

andrewgsavage committed:
Merge remote-tracking branch 'origin/master' into infer_time

# Conflicts:
#	pint/facets/numpy/quantity.py
#	pint/facets/plain/quantity.py

andrewgsavage committed:
Fix Q_(quantity, units) crashing via _is_timedelta's duck-array check

CI failure (docs doctest job): Q_(length, 'cm') raised AttributeError
when length was itself a Quantity, e.g.:

    AttributeError: Neither Quantity object nor its magnitude (30)
    has attribute 'dtype'

Root cause: is_duck_array_type (pint/compat.py) checks
hasattr(cls, "_magnitude")/hasattr(cls, "_units") on the *class* to
rule out Quantity types, but those attributes are only ever set on
*instances* (assigned in __new__), so hasattr on the bare class is
False. Combined with Quantity defining __array_function__/ndim/dtype,
is_duck_array_type(Quantity) misclassifies the class itself as a duck
array. NumpyQuantity._is_timedelta then calls is_duck_array(value) on
a `value` that is itself a Quantity (not yet excluded, since that
check happens later in __new__), gets a false positive, and crashes
accessing value.dtype when the wrapped magnitude has no dtype of its
own (or silently "worked" for ndarray magnitudes only because their
dtype happened not to match timedelta64).

Fix at the call site: exclude PlainQuantity instances from the
duck-array timedelta check, since a Quantity is never itself the kind
of raw timedelta64 array this branch is meant to detect. Leaves the
broader is_duck_array_type heuristic in compat.py untouched, since
fixing that is a separate, wider-reaching change.

andrewgsavage committed:
Update CHANGES to reflect `Quantity` timedelta conversion

Document the new behavior of `Quantity` with `datetime.timedelta` initialization.

andrewgsavage committed:
Update CHANGES for Quantity initialization clarification

Clarified the documentation for initializing a Quantity with datetime.timedelta.
